src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1345 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.FSharp.Collections
+    open System
+    open System.Diagnostics
+    open System.Collections
+    open System.Collections.Generic
+    open System.Reflection
+    open Microsoft.FSharp.Core
+    open Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicOperators
+    open Microsoft.FSharp.Core.Operators
+    open Microsoft.FSharp.Core.CompilerServices
+    open Microsoft.FSharp.Control
+    open Microsoft.FSharp.Collections
+    open Microsoft.FSharp.Primitives.Basics
+    open Microsoft.FSharp.Collections.SeqComposition
+    open Microsoft.FSharp.Collections.SeqComposition.Core
+
+    [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+    module ISeq =
+        open IEnumerator
+
+        module Core =
+            [<Struct; NoComparison; NoEquality>]
+            type Value<'a> =
+                val mutable _1 : 'a
+                new (a:'a) = { _1 = a }
+
+            [<Struct; NoComparison; NoEquality>]
+            type Values<'a,'b> =
+                val mutable _1 : 'a
+                val mutable _2 : 'b
+                new (a:'a, b: 'b) = { _1 = a;  _2 = b }
+
+            [<Struct; NoComparison; NoEquality>]
+            type Values<'a,'b,'c> =
+                val mutable _1 : 'a
+                val mutable _2 : 'b
+                val mutable _3 : 'c
+                new (a:'a, b:'b, c:'c) = { _1 = a; _2 = b; _3 = c }
+
+            [<AbstractClass>]
+            type Transform<'T,'U,'State> =
+                inherit Activity<'T,'U>
+                
+                new (next:Activity, initState:'State) = {
```

Comment:
I do use direct access to the underlying field - which in this case happen to be the 'State, which is a mutable struct, which doesn't work with the implicit constructor. The whole reason for this state hack is that the generated object expressions use individual ref cells for all captured state. In order to minimize work on the GC I have used this state object to do the capturing.